core/utils/syscallthread.h

```diff
@@ -0,0 +1,443 @@
+ *     void Run() { ... code goes here ... }
+ *   };
+ *   class X {
+ *     ...
+ *     SomeThread thread;
+ *     ...
+ *   };
+ *   X var;
+ *   ...
+ *   var.thread.Start();
+ *
+ * and your SomeThread::Run() will be called in the started thread.
+ *
+ * In Run(), check IsExitRequested() any time your system call(s)
+ * return(s) with an EINTR error, as these are requests for your
+ * thread to terminate.  You can check it any other time as well.
+ * (We check it once before even calling Run(), so if you do
+ *
+ * Optionally (for performance / avoiding more signals), you may
+ * call BeginExiting() once you are on the way out, but have not
+ * actually finished.  This tells the knock thread (see below) that
+ * its job is done.
+ *
+ * Once your Run() returns (the thread has exited), var.thread.Done()
+ * will return True.
+ *
+ * You may WaitFor() the thread, or call Terminate(), at any time:
+ * these are no-ops if the thread was never Start()ed.
+ *
+ * Once Start()ed, the thread is not re-Start()able until
+ * Terminate()d and/or WaitFor()ed, after which you may --
+ * CAREFULLY (e.g., under locks if this could race) -- invoke
+ * Reset() to put it back to "never started" state.
+ *
+ * Because many system calls are not available in a reliable-signal
+ * flavor (cf. pselect/ppoll), requesting an exit normally starts a
+ * "knock thread" that keeps kicking your Run() code to get it to
+ * return.
+ *
+ * This part is optional: if you declare that you are using the
+ * reliable signal system call, and never block in any other syscall,
+ * you can skip it.  In this case you should call Sigmask() to
+ * obtain the correct mask to use in pselect/ppol, and then you
+ * *must* check IsExitRequested() immediately after the pselect/ppoll
+ * returns.
+ *
+ * Otherwise, there is a race between any IsExitRequested() test and
+ * the entry to a system call.  This is why we have a knock thread:
+ * it will repeatedly send the interrupt signal.  At least one of
+ * these will cause the system call itself to be interrupted and
+ * return -1 with errno set to EINTR.
+ *
+ * This means that when using SyscallThread<false>, if you need to
+ * make system calls in Run() that must *not* be interrupted, you
+ * should call PushDefer() first.  Knock-thread signals will be
+ * deferred until a corresponding PopDefer().
+ *
+ * TODO(torek): attempt to template away knock_thread_ related code
+ * when using reliable signals.
+ *
+ * MAYBE-to-do(torek): allow detaching (move the state variables
+ * into a sub-object that is given to the thread, so that we can
+ * use thread_.detach() and knock_thread_.detach()).  If we do this
+ * we need new/dispose, std::unique_ptr<...>, and/or refcounts.
+ */
+template <bool ReliableSignals>
+class SyscallThread {
+ public:
+  // NB: order matters here.  Thread state progresses linearly
+  // (except for Reset() which must not be allowed to race).
+  enum ThreadState { TSNotStarted, TSStarting, TSReady, TSExiting, TSDone };
```

Comment:
Style guide says these should be `kFoo` constants: https://google.github.io/styleguide/cppguide.html#Enumerator_Names.  Not sure if TS (short for thread state) should be Ts, or maybe just rely on them being inside the `ThreadState` namespace in the first place and call them `kNotStarted`, `kStarting`, etc.?

Reply:
A new namespace with `enum class` sounds good.

The guide says `Ts` is right. My soul dies whenever I have to type `Tcp`...